### PR TITLE
Feature: Add top level metadata in excel

### LINKED ISSUE #510: [Feature Request]: Swate, Excel, add Metadata
### Contact Details

_No response_

### Is your feature request related to a problem?

Currently no meta data sheet is available for swate-excel.

### Describe the solution you would like!

The sidebar gets a new tab that enables the creation of a meta-data sheet for excel.
This includes the choosing of the meta-data sheets type (Investigation, Study, Assay, or Template) in a drop down menu
The Identifier, Contacts, and Publications, Dates, and so on can all be chosen in the sidebar.

When a meta-data sheet is already available a new one cannot be created but an existing one can be updated.

Normally, the user does not have to interact with the excel sheet at all! (Goal)

### Additional context!

_No response_

### Host

Excel

### Host Details

Excel

### PR DESCRIPTION
**Pull request by Etschbeijer**
closes #510

### CODE REVIEW

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 1]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
...
+
+                return [InteropLogging.Msg.create InteropLogging.Warning $"The work sheet {metadataSheet.name} has been deleted"]
+        }
+    )
+
+/// <summary>
+/// Converts a sequence of sequence of excel data into a resizearray, compatible with Excel.Range
+/// </summary>
+/// <param name="metadataValues"></param>
+let private convertToResizeArrays metadataValues =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/OfficeInterop/OfficeInterop.fs:2172]: Plus i am not sure what this is meant to do. It looks extremely confusing.

If it is meant to fill missing values to keep all inner sequences on the same length. It appears to be buggy:

```fsharp
let example: obj option seq seq = seq [
  seq [Some 1; Some 2; None; Some 3];
  seq [Some 4; Some 5];
  seq [Some 6; Some 7; Some 8; Some 9]
]

convertToResizeArrays example

// seq
//    [seq [Some 1; Some 2; None; Some 3]; 
//     seq [Some 4; Some 5; None]; //only 3 values here
//     seq [Some 6; Some 7; Some 8; Some 9]]
```

Check if this appending is really necessary and i yes you might want to consider the function below:

```
let convertToResizeArrays2 (metadataValues: seq<seq<obj option>>) =
  let ra = ResizeArray()
  let maxLength = 
    example |> Seq.fold (fun acc ele ->
      let l = Seq.length ele
      if l > acc then l else acc
    ) 0
  for seq in metadataValues do
    let ira = ResizeArray seq
    if ira.Count < maxLength then
      ira.AddRange (Array.create (maxLength - ira.Count) None)
    ra.Add (ira)
  ra
```

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 2]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
...
+
+open FsSpreadsheet
+
+/// <summary>
+/// Deletes the data contained in the selected worksheet and fills it afterwards with the given new data
+/// </summary>
+/// <param name="context"></param>
+/// <param name="fsWorkSheet"></param>
+/// <param name="seqOfSeqs"></param>
+let private updateWorkSheet (context:RequestContext) (fsWorkSheet:FsWorksheet) (seqOfSeqs:seq<seq<string option>>) =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2192]: This function does not delete all data in a worksheet only all data in the range of the new data

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 3]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
...
+    )
+
+/// <summary>
+/// Delete excel worksheet that contains top level metadata
+/// </summary>
+/// <param name="identifier"></param>
+let deleteTopLevelMetadata (identifier: string option) =
+    Excel.run(fun context ->
+        promise {
+            if identifier.IsNone then
```
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:2147]: Catching a potential error that should not occur. Added that part to the message

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 4]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
...
+                assayWorkSheet.load(propertyNames = U2.Case2 (ResizeArray[|"name"|])) |> ignore
+                range.load(propertyNames = U2.Case2 (ResizeArray["values"]))
+
+            do! context.sync().``then``(fun _ -> ())
+
+            let values =
+                range.values
+                |> Seq.map (fun x ->
+                    x
+                    |> Seq.map (fun xx -> if xx.IsSome && not (String.IsNullOrEmpty(xx.Value.ToString())) then Some (xx.Value.ToString()) else None))
```
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 5]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
         }
     )
 
-// Old stuff, mostly deprecated
+/// <summary>
+/// Tries to get the name of the top level excel worksheet of top level metadata
+/// </summary>
+let tryGetTopLevelMetadataSheetName () =
```
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 2 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:2195]: It should delete all data beause the first range accessed is:
let range = worksheet.getUsedRange true
This selects all cells that contain a value or are formatted and those are set to null.

In the next step, the cells are selected, that are in the range of the new data.

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 6]
```diff
@@ -2075,7 +2064,255 @@ let rectifyTermColumns () =
...
+/// </summary>
+/// <param name="name"></param>
+let createTopLevelMetadata name =
+    Excel.run(fun context ->
+        promise {
+            let workSheetName = $"{name}"
+            let newWorkSheet = context.workbook.worksheets.add(workSheetName)
+            newWorkSheet.activate()
+            do! context.sync().``then``(fun _ -> ())
+            let! result = tryGetTopLevelMetadataSheetName()
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: was it ever not successful without excel trhowing an error?

File: src/Client/Client.fsproj [hunk 7]
```diff
@@ -66,6 +66,13 @@
     <Compile Include="SharedComponents\AdvancedSearch.fs" />
     <Compile Include="SharedComponents\TermSearchInput.fs" />
     <Compile Include="SharedComponents\QuickAccessButton.fs" />
+    <Compile Include="SharedComponents\Generic.fs" />
```
Comment (Freymaurer) [inline src/Client/Client.fsproj:69]: What exactly did you do here? The files in maincomponents still exist?

File: src/Client/Update/OfficeInteropUpdate.fs [hunk 8]
```diff
@@ -221,22 +221,70 @@ module OfficeInterop =
...
+                state, model, cmd
+            | UpdateTopLevelStudy study ->
+                let cmd =
+                    Cmd.OfPromise.either
+                        OfficeInterop.Core.updateTopLevelStudy
+                        (study)
+                        (curry GenericInteropLogs Cmd.none >> DevMsg)
+                        (curry GenericError Cmd.none >> DevMsg) //error
+                state, model, cmd
+            | UpdateTopLevelTemplate template ->
```
Comment (Freymaurer) [inline src/Client/Update/OfficeInteropUpdate.fs:256]: make these 4 one function using arcfiles

File: src/Client/SidebarComponents/Navbar.fs [hunk 9]
```diff
@@ -1,25 +1,374 @@
...
+    static member init = {
+        MetadataType = None
+        Identifier = None
+        Assay = None
+        Investigation = None
+        Study = None
+        Template = None
+    }
 
+type WorkSheetNames =
```
Comment (Freymaurer) [inline src/Client/SidebarComponents/Navbar.fs:51]: this does not work with backwards compatibility use arctrl function

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 10]
```diff
@@ -172,6 +172,26 @@ module OfficeInteropExtensions =
...
+
+            //Selects the longest sequence of the metadata values
+            //In the next step, determines the length of the longest metadata value sequence
+            let maxLength = metadataValues |> Seq.maxBy Seq.length |> Seq.length
+
+            //Adapts the length of the smaller sequences to the length of the longest sequence in order to avoid problems with the insertion into excel.Range
+            let ra = ResizeArray()
+            for seq in metadataValues do
+                //Parse string to obj option
+                let ira = ResizeArray (seq |> Seq.map (fun column -> if column.IsSome then Some (column.Value :> obj) else None))
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:189]: Option.map

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 11]
```diff
@@ -2075,7 +2084,237 @@ let rectifyTermColumns () =
...
+            else
+                return [InteropLogging.Msg.create InteropLogging.Error $"No study is available"]
+        }
+    )
+
+/// <summary>
+/// Updates top level metadata excel worksheet of templates
+/// </summary>
+/// <param name="assay"></param>
+let updateTopLevelTemplate (template: Template option) =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2299]: make these 4 one function using arcfiles

File: src/Client/SidebarComponents/Navbar.fs [hunk 12]
```diff
@@ -1,25 +1,374 @@
...
+                                    excelMetadataType with
+                                        MetadataType = Some ArcFilesDiscriminate.Template
+                                        Identifier = Some templateIdentifier
+                                        Template = template
+                                }
+                            | _ -> failwith $"No metadata of type {result.Value} has been implemented yet!"
+                        else setExcelMetadataType(ExcelMetadataState.init)
+                    }
+                )  |> ignore
+            )
```
Comment (Freymaurer) [inline src/Client/SidebarComponents/Navbar.fs:369]: can show a gif of this ui?

i can recommend screentogif as tool

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/Client.fsproj (hunk 7 above)
Comment (Etschbeijer) [reply src/Client/Client.fsproj:69]: They still exist. I wanted to remove them in an extra Issue.

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: Not yet but for the case I wanted to be prepared.

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 11 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 10 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/SidebarComponents/Navbar.fs (hunk 9 above)
Comment (Etschbeijer) [reply src/Client/SidebarComponents/Navbar.fs]: fixed

File: src/Client/Update/OfficeInteropUpdate.fs (hunk 8 above)
Comment (Etschbeijer) [reply src/Client/Update/OfficeInteropUpdate.fs]: fixed

File: src/Client/SidebarComponents/Navbar.fs (hunk 12 above)
Comment (Etschbeijer) [reply src/Client/SidebarComponents/Navbar.fs:335]: ExcelSwateTopLevelMetadataShowcase18102024PB
thanks for the tool suggestion! It is really cool, I hope this is helpful

File: src/Client/SidebarComponents/Navbar.fs [hunk 13]
```diff
@@ -67,11 +51,10 @@ let createMetaDataTypeButtons excelMetadataType setExcelMetadataType (dispatch:
...
                     setExcelMetadataType {
                         excelMetadataType with
-                            Identifier = Some WorkSheetNames.investigation
-                            MetadataType = Some ArcFilesDiscriminate.Investigation
-                            Investigation = Some investigation
+                            WorkSheetName = Some ArcInvestigation.metadataSheetName
+                            TopLevelMetadata = Some (ArcFiles.Investigation investigation)
                     }
-                    OfficeInterop.CreateTopLevelMetadata(WorkSheetNames.investigation)
+                    OfficeInterop.CreateTopLevelMetadata(ArcInvestigation.metadataSheetName)
```
Comment (Freymaurer) [inline src/Client/SidebarComponents/Navbar.fs:57]: this is still not backwards compatible. You must use the isMetadataSheet function

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/SidebarComponents/Navbar.fs (hunk 13 above)
Comment (Etschbeijer) [reply src/Client/SidebarComponents/Navbar.fs:57]: fixed

File: src/Client/Client.fsproj (hunk 7 above)
Comment (Etschbeijer) [reply src/Client/Client.fsproj:69]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 14]
```diff
@@ -2075,7 +2086,179 @@ let rectifyTermColumns () =
...
+
+        let _ = worksheets.load(propertyNames = U2.Case2 (ResizeArray[|"values"; "name"|]))
+
+        do! context.sync().``then``(fun _ -> ())
+
+        let worksheetItems = worksheets.items |> Array.ofSeq
+
+        let metadata =
+            worksheetItems
+            |> Array.choose (fun item -> if item.name.Contains("isa_") then Some item else None)
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2104]: use ARCtrl logic to check for metadatasheets

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 15]
```diff
@@ -2075,7 +2086,179 @@ let rectifyTermColumns () =
...
+
+        let potMetadata =
+            if metadata.Length < 1 then None
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2107]: match metadata with
| Some m -> m.[0].name
| None -> failwith ...

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 16]
```diff
@@ -2075,7 +2086,179 @@ let rectifyTermColumns () =
...
+let deleteTopLevelMetadata () =
+    Excel.run(fun context ->
+        promise {
+            let worksheets = context.workbook.worksheets
+
+            let _ = worksheets.load(propertyNames = U2.Case2 (ResizeArray[|"values"; "name"|]))
+
+            do! context.sync().``then``(fun _ -> ())              
+
+            worksheets.items
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2177]: very nice!

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 17]
```diff
@@ -2075,7 +2086,179 @@ let rectifyTermColumns () =
...
+/// </summary>
+/// <param name="name"></param>
+let createTopLevelMetadata name =
+    Excel.run(fun context ->
+        promise {
+            let workSheetName = $"{name}"
+            let newWorkSheet = context.workbook.worksheets.add(workSheetName)
+            newWorkSheet.activate()
+            do! context.sync().``then``(fun _ -> ())
+            let! result = tryGetTopLevelMetadataSheetName context
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:2125]: remove this check and instead wrap the whole code block starting with `newWorkSheet.activate()` in a try ... with

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 16 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:2177]: Thank you!

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 14 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 15 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 17 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

**Review by Freymaurer — APPROVED** [review]

**Pull request merged**